EarthModelBinned.h

```diff
@@ -0,0 +1,171 @@
+    double zoa;     ///< The effective Z/A value of the matter in the bin
+    int index;      ///< Region index
+
+  };
+
+  class EarthModelBinned : public EarthModelBase {
+
+    public:
+
+      EarthModelBinned(std::string filename=""); ///< Constructor
+      virtual ~EarthModelBinned();          ///< Destructor
+
+      int FillPath(double cosT, double phi=0); ///< Fill the path sequence in a vector (phi in degrees)
+
+//      virtual void SetDetPos(double dep, double lat, double lon); ///< Set the detector position (dep = depth in km, lat/lon in deg)
+
+      virtual void LoadModel(std::string filename); ///< Load an earth model from a file
+
+      virtual std::vector<OscProb::EarthBin> GetEarthBins(); ///< Get the set of earth layers
+
+      virtual void SetRegionZoA(int index, double zoa); ///< Set Z/A of all bins with specified region index
+      virtual double GetRegionZoA(int index); ///< Get Z/A of all bins with specified region index
+      virtual void ScaleRegionDensity(int index, double scalingfactor); ///< Set Z/A of all bins with specified region index
+
+    protected:
+
+      virtual void ClearModel(); ///< Clear the earth model information
+
+      virtual void AddBin(double radius_out, double radius_in, double latitude, double longitude, double density, double zoa, double layer); ///< Add a bin to the model (angles in degrees)
+
+      virtual void AddPath(double length, EarthBin bin);  ///< Add a path segment to the sequence
+
+      virtual double DetDistForNextLatBin(int cur_index, int &dir, double dLat, double maxSinSqLat, int &max_reached, double beta, double gamma, double gammaSq, double rDetGammaSinDetLat, double rDetCosAcosDetLat, double rDetSinDetLat, double rDetCosT, double rDetSinT); ///< Calculate the detector distance at the edge of the current lat bin along the neutrino's trajectory
+
+      virtual double DetDistForNextLonBin(double prev_lon, double dLon, int &lon_bin, double min_lon, double max_lon, double sinDetLon, double cosDetLon, double alpha, double sinTsinAsinDetLon, double sinTsinAcosDetLon, double rDetCosDetLat); ///< Calculate the detector distance at the edge of the current lon bin along the neutrino's trajectory and increment lon_bin
+
+      virtual void RecordLatLonBinCrossings(double detDist_nextDbin, double &DetDist, int &index, int &latBin, int &nextLatBin, double &detDist_nextLatBin, int &sign, double dLat, double maxSinSqLat, int &maxlatreached, int &lonBin, int &nextLonBin, double &detDist_nextLonBin, double dLon, double min_lon, double max_lon, double alpha, double beta, double gamma, double gammaSq, double rDetGammaSinDetLat, double rDetCosAcosDetLat, double rDetSinDetLat, double rDetCosDetLat, double rDetCosT, double rDetSinT, double sinTsinAsinDetLon, double sinTsinAcosDetLon, double sinDetLon, double cosDetLon); ///< Record path segments for each latitude/longitude bin crossed before reaching detDist_nextDbin
```

Comment:
Many of these functions have too many arguments. Especially in C++, this is not great because you need them to be input in the exact order they were declared. Instead you should create some classes or structs to represent the data you want to pass in or out.

Reply:
Many of the arguments for DetDistForNextLatBin, DetDistForNextLonBin, and RecordLatLonBinCrossings have now been replaced by arguments of structs LatBinInfo and LonBinInfo.  I also created a struct (TrajConstants) to hold most of the variables that are different combinations of the detector coordinates and neutrino trajectory angles.  An private object of that struct is included in the EarthModelBinned class, so those variables are no longer included as arguments of functions.  Does that address all your concerns about this?

Reply:
Looks, much better. If all tests pass I will merge this pull request. Thanks a lot!